### PR TITLE
Refactor TypeRegistry

### PR DESCRIPTION
**Pull request by arhimondr**

### CODE REVIEW

**Comment by arhimondr** [comment]
@losipiuk @martint Please review. I can merge it to https://github.com/prestodb/presto/pull/4958 patch if looks good.

File: presto-blackhole/src/main/java/com/facebook/presto/plugin/blackhole/BlackHoleConnector.java [hunk 1]
```diff
@@ -24,6 +24,7 @@
 import com.facebook.presto.spi.session.PropertyMetadata;
 import com.facebook.presto.spi.transaction.IsolationLevel;
 import com.facebook.presto.spi.type.TypeManager;
+import com.facebook.presto.spi.type.TypeSignatureParameter;
```
Comment (losipiuk) [inline presto-blackhole/src/main/java/com/facebook/presto/plugin/blackhole/BlackHoleConnector.java:27]: Extend commit message to explain what is done and why. And change TypeRegistry to TypeManager in commit message.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java [hunk 2]
```diff
@@ -191,6 +193,11 @@ public Type getCoercion(Expression expression)
         return coercions.get(expression);
     }
 
+    public boolean isTypeOnlyCoercion(Expression expression)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: changes in this file seem unrelated to commit message

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalysis.java [hunk 3]
```diff
@@ -61,6 +64,11 @@ public Type getCoercion(Expression expression)
         return expressionCoercions.get(expression);
     }
 
+    public boolean isTypeOnlyCoercion(Expression expression)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalysis.java:67]: ditto

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 4]
```diff
@@ -1287,14 +1285,14 @@ protected RelationType visitValues(Values node, AnalysisContext context)
                     Type expectedType = fieldTypes.get(i);
                     Expression item = items.get(i);
                     if (!analysis.getType(item).equals(expectedType)) {
-                        analysis.addCoercion(item, expectedType);
+                        analysis.addCoercion(item, expectedType, false);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: always false?

File: presto-main/src/test/java/com/facebook/presto/type/TestTypeRegistry.java [hunk 5]
```diff
@@ -29,109 +27,90 @@
...
 import static org.testng.Assert.assertEquals;
 import static org.testng.Assert.assertFalse;
 import static org.testng.Assert.assertTrue;
 
 public class TestTypeRegistry
 {
-    private final TypeSignature varcharX = new TypeSignature(StandardTypes.VARCHAR, ImmutableList.of(TypeSignatureParameter.of("x")));
-    private final TypeSignature decimalPS = new TypeSignature(StandardTypes.DECIMAL,
-            ImmutableList.of(TypeSignatureParameter.of("p"), TypeSignatureParameter.of("n")));
+    private static final TypeRegistry TYPE_REGISTRY = new TypeRegistry();
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/type/TestTypeRegistry.java]: make this non-static.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 6]
```diff
@@ -227,62 +228,47 @@ private static boolean isCovariantParameterPosition(String firstTypeBase, int po
...
+            boolean sourcePrecisionIsLessOrEqualToResultPrecision = sourceDecimal.getPrecision() <= resultDecimal.getPrecision();
+            return sameDecimalSubtype && sameScale && sourcePrecisionIsLessOrEqualToResultPrecision;
+        }
 
-                TypeSignatureParameter actualParameter = actualType.getParameters().get(i);
-                TypeSignatureParameter expectedParameter = expectedType.getParameters().get(i);
-                if (actualParameter.equals(expectedParameter)) {
-                    continue;
-                }
+        String sourceTypeName = source.getTypeSignature().getBase();
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: Maybe call this sourceTypeBase and resultTypeBase to be consistent with getter name.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 7]
```diff
@@ -227,62 +228,47 @@ private static boolean isCovariantParameterPosition(String firstTypeBase, int po
...
+        String resultTypeName = result.getTypeSignature().getBase();
 
-                Optional<TypeSignature> actualParameterSignature = actualParameter.getTypeSignatureOrNamedTypeSignature();
-                Optional<TypeSignature> expectedParameterSignature = expectedParameter.getTypeSignatureOrNamedTypeSignature();
-                if (!actualParameterSignature.isPresent() || !expectedParameterSignature.isPresent()) {
+        if (sourceTypeName.equals(resultTypeName)) {
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: this is true for ARRAY, MAP and ROW. But it is not general truth. We can leave it for now. As we do not have more types taking types as parameters.

File: presto-kafka/src/test/java/com/facebook/presto/kafka/TestKafkaPlugin.java [hunk 8]
```diff
@@ -104,6 +104,12 @@ public boolean isTypeOnlyCoercion(Type actualType, Type expectedType)
         {
             return Optional.empty();
         }
+
+        @Override
+        public Optional<Type> coerceTypeBase(Type sourceType, String resultTypeName)
```
Comment (losipiuk) [inline presto-kafka/src/test/java/com/facebook/presto/kafka/TestKafkaPlugin.java]: s/resultTypeName/resultTypeBase/

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 9]
```diff
@@ -273,287 +213,224 @@ public boolean isTypeOnlyCoercion(Type source, Type result)
...
+                    case StandardTypes.INTERVAL_DAY_TO_SECOND:
+                    case RegexpType.NAME:
+                    case LikePatternType.NAME:
+                    case JsonPathType.NAME:
+                    case ColorType.NAME:
+                        return Optional.of(getType(new TypeSignature(resultTypeName)));
+                    case StandardTypes.VARCHAR:
+                        return Optional.of(createVarcharType(0));
+                    case StandardTypes.DECIMAL:
+                        return Optional.of(createDecimalType(1, 0));
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: It should coerce to `DECIMAL(0,0)`. 
Otherwise we end up with situation that `CAST(NULL as DECIMAL(x,y))` would not coerce to `DECIMAL(1,1)`.

It is problematic as we disallow DECIMAL(0,0) right now.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java [hunk 10]
```diff
@@ -273,287 +213,224 @@ public boolean isTypeOnlyCoercion(Type source, Type result)
...
+                    case StandardTypes.ARRAY:
+                        return Optional.of(new ArrayType(UnknownType.UNKNOWN));
+                    case StandardTypes.MAP:
+                        return Optional.of(new MapType(UnknownType.UNKNOWN, UnknownType.UNKNOWN));
+                    default:
+                        return Optional.empty();
+                }
+            }
+            case StandardTypes.INTEGER: {
+                switch (resultTypeName) {
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: decimal missing

Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: As we spoke.
Eventually adde validator which checks that 
`if there is coercion A->B and coercion B->C
that there is coercion A->C`

```
```

**Comment by losipiuk** [comment]
generally LGTM; see the comments

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java (hunk 2 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/Analysis.java]: It is really hard to separate those 2 commits now. And actually keeping those changes in separate is confusing. So it is hard to understand why those changes are being made. Instead of separating this commit into 2 I'm just going to change the commit message.

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalysis.java (hunk 3 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/ExpressionAnalysis.java:67]: ditto

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java (hunk 4 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: Nice catch. I temporarily changed it to `false` just to make the tests pass, and than i forgot about this place.

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java (hunk 10 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: I'm going to introduce such test together with the patch that will introduce `INTEGER` to `DECIMAL` coercion

File: presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java (hunk 9 above)
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java]: Creation of the `DECIMAL(0,0)` type is not allowed right now. I think it should be changed in a separate patch.

File: presto-main/src/test/java/com/facebook/presto/type/TestTypeRegistry.java [hunk 11]
```diff
@@ -172,6 +179,57 @@ public void testCastTypeBaseParametric()
...
+            }
+        }
+    }
+
+    private Set<Type> getStandardPrimitiveTypes()
+    {
+        ImmutableSet.Builder<Type> builder = ImmutableSet.builder();
+        // add unparametrized types
+        builder.addAll(typeRegistry.getTypes());
+        builder.add(createDecimalType(1, 0));
```
Comment (losipiuk) [inline presto-main/src/test/java/com/facebook/presto/type/TestTypeRegistry.java]: this seems a bit arbitrary.

**Comment by arhimondr** [comment]
Merged to master
